tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+
+  unsigned char *received_nonce_a = NULL;
+  size_t received_nonce_a_len = 0;
+
+  unsigned char *received_id = NULL;
+  size_t received_id_len = 0;
+
+  result = parse_nonce_response(private_key_ctx,
+                                response, read_result,
+                                &received_nonce_a, &received_nonce_a_len,
+                                &nonce_b, &nonce_b_len,
+                                &received_id, &received_id_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to parse the nonce response.");
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Received nonce A: %zd bytes", received_nonce_a_len);
+  kmyth_log(LOG_INFO, "Received nonce B: %zd bytes", nonce_b_len);
+  kmyth_log(LOG_INFO, "Received ID: %.*s", received_id_len, received_id);
+
+  if (nonce_a_len != received_nonce_a_len)
+  {
+    kmyth_log(LOG_ERR, "The received nonce A length is invalid.");
+    return 1;
+  }
+  if (strncmp
+      ((const char *) nonce_a, (const char *) received_nonce_a,
+       nonce_a_len) != 0)
+  {
+    kmyth_log(LOG_ERR, "The received nonce A is invalid.");
+    kmyth_log(LOG_ERR, "Expected nonce A: %zd bytes", nonce_a_len);
+    return 1;
+  }
+
+  if (strncmp
+      ((const char *) received_id, (const char *) expected_id,
+       expected_id_len) != 0)
+  {
+    kmyth_log(LOG_ERR, "The received ID is invalid.");
+    return 1;
+  }
+
+  result = build_nonce_confirmation(public_key_ctx,
+                                    nonce_b, nonce_b_len,
+                                    &request, &request_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to build the nonce confirmation.");
+    return 1;
+  }
+
+  if (write(socket_fd, request, request_len) != request_len)
+  {
+    kmyth_log(LOG_ERR, "Failed to fully send the nonce confirmation.");
+    return 1;
+  }
+
+  free(request);
+  request = NULL;
+  request_len = 0;
+
+  // Use nonces to generate shared session key S
+  result = generate_session_key(nonce_a, nonce_a_len,
+                                nonce_b, nonce_b_len,
+                                session_key, session_key_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to generate the session key.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// negotiate_server_session_key()
+//
+int negotiate_server_session_key(int socket_fd,
+                                 EVP_PKEY_CTX * public_key_ctx,
+                                 EVP_PKEY_CTX * private_key_ctx,
+                                 unsigned char *id, size_t id_len,
+                                 unsigned char **session_key,
+                                 size_t *session_key_len)
+{
+  // Generate nonce B
+  unsigned char *nonce_b = NULL;
+  size_t nonce_b_len = 0;
+
+  int result = generate_nonce(32, &nonce_b, &nonce_b_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to generate a nonce.");
+    return 1;
+  }
+
+  // Conduct NSL to obtain nonce A
+  unsigned char *response = calloc(8192, sizeof(unsigned char));
```

Comment:
We should check that the return value of calloc() is not NULL.

Reply:
Done